services/horizon/internal/sse_test.go

```diff
@@ -0,0 +1,194 @@
+	defer ht.Finish()
+
+	ht.App.ticks.Stop()
+
+	channelA := sse.Subscribe("a")
+	channelB := sse.Subscribe("b")
+	defer sse.Unsubscribe(channelA, "a")
+	defer sse.Unsubscribe(channelB, "b")
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(channelA chan interface{}, channelB chan interface{}, wg *sync.WaitGroup) {
+		defer wg.Done()
+		select {
+		case <-channelA:
+			select {
+			case <-channelA:
+				t.Error("channelA shouldnt trigger")
+				t.FailNow()
+			case <-channelB:
+				t.Error("channelB shouldnt trigger")
+				t.FailNow()
+			case <-time.After(2 * time.Second):
+				// no-op. Success!
+			}
+		case <-channelB:
+			t.Error("channelB shouldnt trigger")
+			t.FailNow()
+		case <-time.After(2 * time.Second):
+			t.Error("channelA did not trigger")
+			t.FailNow()
+		}
+	}(channelA, channelB, &wg)
+
+	sse.Publish("a")
+	wg.Wait()
+}
+
+// Check multiple number of topics handled
+func TestSSEPubsubManyChannels(t *testing.T) {
+	ht := StartHTTPTest(t, "base")
+	defer ht.Finish()
+
+	ht.App.ticks.Stop()
+
+	var wg sync.WaitGroup
+	wg.Add(100)
+	subscriptions := make([]chan interface{}, 100)
+	for i := 0; i < 100; i++ {
+		subscriptions[i] = sse.Subscribe(strconv.Itoa(i))
+		defer sse.Unsubscribe(subscriptions[i], strconv.Itoa(i))
+		go func(subscription chan interface{}, wg *sync.WaitGroup) {
+			defer wg.Done()
+			select {
+			case <-subscription:
+				return
+			case <-time.After(2 * time.Second):
+				t.Error("Subscription did not trigger within 2 seconds")
+				t.FailNow()
+			}
+		}(subscriptions[i], &wg)
+	}
+
+	for j := 0; j < 100; j++ {
+		sse.Publish(strconv.Itoa(j))
+	}
+
+	wg.Wait()
+}
+
+// Check multiple subscriptions to the same topic.
+func TestSSEPubsubManyListeners(t *testing.T) {
+	ht := StartHTTPTest(t, "base")
+	defer ht.Finish()
+
+	ht.App.ticks.Stop()
+
+	var wg sync.WaitGroup
+	wg.Add(100)
+	subscriptions := make([]chan interface{}, 100)
+	//var subscriptions [100]chan interface{}
+	for i := 0; i < 100; i++ {
+		subscriptions[i] = sse.Subscribe("a")
+		defer sse.Unsubscribe(subscriptions[i], "a")
+		go func(subscription chan interface{}, wg *sync.WaitGroup) {
+			defer wg.Done()
+			select {
+			case <-subscription:
+				return
+			case <-time.After(2 * time.Second):
+				t.Error("Subscription did not trigger within 2 seconds")
+				t.FailNow()
+			}
+		}(subscriptions[i], &wg)
+	}
+
+	sse.Publish("a")
+
+	wg.Wait()
+}
+
+// Check sse subscription get message when ingest to Horizon happens.
+func TestSSEPubsubTransactions(t *testing.T) {
+	tt := test.Start(t).ScenarioWithoutHorizon("kahuna")
+	defer tt.Finish()
+	subscription := sse.Subscribe("GA46VRKBCLI2X6DXLX7AIEVRFLH3UA7XBE3NGNP6O74HQ5LXHMGTV2JB")
+	defer sse.Unsubscribe(subscription, "GA46VRKBCLI2X6DXLX7AIEVRFLH3UA7XBE3NGNP6O74HQ5LXHMGTV2JB")
+	var wg sync.WaitGroup
+	wg.Add(10)
+
+	go func(subscription chan interface{}, wg *sync.WaitGroup) {
+		for i := 0; i < 10; i++ {
+			select {
+			case <-subscription:
+				wg.Done()
+			case <-time.After(10 * time.Second):
```

Comment:
bug: if the time.After returns before subscription the wait at the bottom will hang indefinitely